src/Entity/PathautoPattern.php

```diff
@@ -0,0 +1,384 @@
+    $contexts = $this->getContexts();
+    foreach ($this->getAliasType()->getContexts() as $plugin_context_id => $plugin_context) {
+      unset($contexts[$plugin_context_id]);
+    }
+    $this->context_definitions = [];
+    foreach ($contexts as $context_id => $context) {
+      $this->context_definitions[] = [
+        'id' => $context_id,
+        'label' => $context->getContextDefinition()->getLabel()
+      ];
+    }
+  }
+
+  public static function postLoad(EntityStorageInterface $storage, array &$entities) {
+    /** @var \Drupal\ctools\TypedDataResolver $resolver */
+    $resolver = \Drupal::service('ctools.typed_data.resolver');
+    /** @var \Drupal\pathauto\Entity\PathautoPattern $entity */
+    foreach ($entities as $entity) {
+      foreach ($entity->getContextDefinitions() as $definition) {
+        $context = $resolver->convertTokenToContext($definition['id'], $entity->getContexts());
+        $new_definition = new ContextDefinition(
+          $context->getContextDefinition()->getDataType(),
+          $definition['label'],
+          $context->getContextDefinition()->isRequired(),
+          $context->getContextDefinition()->isMultiple(),
+          $context->getContextDefinition()->getDescription(),
+          $context->getContextDefinition()->getDefaultValue()
+        );
+        $new_context = new Context($new_definition, $context->hasContextValue() ? $context->getContextValue() : NULL);
+        $entity->addContext($definition['id'], $new_context);
+      }
+    }
+    parent::postLoad($storage, $entities);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function calculateDependencies() {
+    parent::calculateDependencies();
+
+    // @todo get dependencies from the alias type plugin.
+
+    foreach ($this->getSelectionConditions() as $instance) {
+      $this->calculatePluginDependencies($instance);
+    }
+
+    return $this->getDependencies();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getPattern() {
+    return $this->pattern;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setPattern($pattern) {
+    $this->pattern = $pattern;
+    $this->getAliasType()->setConfiguration(['default' => $pattern]);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getType() {
+    return $this->type;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAliasType() {
+    if (!$this->aliasTypeCollection) {
+      $this->aliasTypeCollection = new DefaultSingleLazyPluginCollection(\Drupal::service('plugin.manager.alias_type'), $this->getType(), ['default' => $this->getPattern()]);
+    }
+    return $this->aliasTypeCollection->get($this->getType());
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getWeight() {
+    return $this->weight;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setWeight($weight) {
+    $this->weight = $weight;
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function hasContext($token) {
```

Comment:
would be nice to have a trait in ctools or so for all those context methods, they're pretty much the same in everything that has context, right?